tools/codegen/gen.py

```diff
@@ -75,20 +77,28 @@ def construct_mapping(self, node, deep=False):  # type: ignore[no-untyped-def]
         mapping['__line__'] = node.start_mark.line + 1
         return mapping
 
-# Parse native_functions.yaml into a sequence of NativeFunctions
-def parse_native_yaml(path: str) -> List[NativeFunction]:
+# Parse native_functions.yaml into a sequence of NativeFunctions and Backend Indices.
+def parse_native_yaml(path: str) -> Tuple[List[NativeFunction], Dict[DispatchKey, BackendIndex]]:
     with open(path, 'r') as f:
         es = yaml.load(f, Loader=LineLoader)
     assert isinstance(es, list)
     rs: List[NativeFunction] = []
+    bs: Dict[DispatchKey, Dict[OperatorName, BackendMetadata]] = defaultdict(dict)
     for e in es:
         assert isinstance(e.get('__line__'), int), e
         loc = Location(path, e['__line__'])
         funcs = e.get('func')
         with context(f'in {loc}:\n  {funcs}'):
-            rs.append(NativeFunction.from_yaml(e, loc))
+            func, m = NativeFunction.from_yaml(e, loc)
+            rs.append(func)
+            BackendIndex.grow_index(bs, m)
     error_check_native_functions(rs)
-    return rs
+    # Default dict is to prevent the codegen from barfing when we have a dispatch key that has no kernels yet.
+    indices: Dict[DispatchKey, BackendIndex] = defaultdict(lambda: BackendIndex(DispatchKey.Undefined, True, {}))
```

Comment:
Might be worth having this constructor on BackendIndex itself. It's a little weird though; the DispatchKey for these fake entries is kind of wrong, so I'm not sure this is a true "zero" element. What's the True? Reduce boolean blindness by kwarg-ing

Reply:
Yeah, kwarging will help. I didn't really want to add this, but there's a new dispatch key that was added recently, that has no kernels associated with it - `SparseCsrCPU`/`SparseCsrCUDA` - which causes the codegen to barf unless we either (a) add at least one native kernel with that backend, which hasn't happened yet, or (b) deal with this case.

Reply:
Can we make it deal? No kernels seems like a case that should work.

Reply:
A cleaner way to make this work without a defaultdict would probably just be to loop through all DispatchKeys and add a corresponding BackendIndex entry if one doesn't exist already- that will at least let us set the `dispatch_key` correctly.

Although it would be kind of nice to have each backend only get added once, rather than XLA getting added as an empty entry during `parse_native_yaml` and later updated in `gen_backend_stubs`.

Maybe a better way to deal with that would be to hardcode which of the dispatch keys are actually external vs. internal, and only generate defaults for internal ones, but that doesn't seem like it provides that much benefit.